services/process-dependencies

```diff
@@ -0,0 +1,227 @@
+    return processManifest(response);
+  } catch (err) {
+    if (err.statusCode == 404) {
+      logger.error(`Could not find ${pluginUrl}`);
+    }
+    else {
+      logger.warn(err);
+    }
+    return null;
+  }
+};
+
+/*
+ * processPlugin will generate the appropriate ingest metadata for the given
+ * plugin based on its version
+ */
+const processPlugin = (plugin, into) => {
+  logger.info('Processing', plugin.artifactId);
+
+  const pluginFilename = path.join(plugin.artifactId, plugin.version, `${plugin.artifactId}-${plugin.version}.hpi`);
+  const groupPath = plugin.groupId.replace(/\./g, '/');
+  let url = `${RELEASES}${groupPath}/`;
+
+  if (plugin.version.match(/(.*?)-rc(\d+)\.(.*)?/)) {
+    logger.debug(`${plugin.artifactId} looks incremental`);
+    url = `${INCREMENTALS}${groupPath}/`;
+  }
+  url = url + pluginFilename;
+
+  let record = {
+    groupId: plugin.groupId,
+    artifactId: plugin.artifactId,
+    url: url,
+    checksum: {},
+  };
+
+  into.push(record);
+
+  tasks.push(
+    request({
+        method: 'HEAD',
+        uri: url,
+        resolveWithFullResponse: true,
+    }).then((response) => {
+      Object.assign(record.checksum, {
+          type: 'sha256',
+          signature: response.headers['x-checksum-sha256'],
+      });
+    })
+  );
+
+  return fetchPluginManifest(url).then(d => d);
+};
+
+// The plugin name map will help us later determine what has been specified in
+// the ssentials.yaml compared to an implied dependency
+const pluginNameMap = new Map(essentials.spec.plugins.map(entry => [entry.artifactId, entry]));
+// Perform a first pass on the plugins needed for all Evergreen installations
+const firstPass = essentials.spec.plugins.map(plugin => processPlugin(plugin, ingest.plugins));
+
+// Collection of plugins which are needed but not specified in essentials.yaml
+let needed = {};
+
+
+const resolvePlugin = (name, inventory) => {
+  // We don't have the plugin in essentials.yaml, so it's dependencies
+  // should be resolved through the update center unless otherwise
+  // specified
+  if (!pluginNameMap[name]) {
+    let releasedPlugin = updates.plugins[name];
+    inventory[name] = releasedPlugin.version;
+    let found = [];
+    found.concat(releasedPlugin.dependencies
+      .filter(d => (d.optional == false))
+      .forEach(d => resolvePlugin(d.name, inventory)));
+    return found;
+  }
+  return [];
+};
+
+const filterPluginsDependencies = (plugins, inventory) => {
+  plugins
+    .filter(d => d)
+    .forEach((depends) => {
+      depends.filter(d => (d.length == 2)).forEach((depend) => {
+        let [name, version] = depend;
+        resolvePlugin(name, inventory);
+      });
+    });
+  return inventory;
+};
+
+Promise.all(firstPass).then(response => filterPluginsDependencies(response, needed))
+.then((inventory) => {
+  Object.keys(inventory).forEach((name) => {
+    ingest.plugins.push({
+        groupId: 'unknown',
```

Comment:
Right…if the `groupId` is used by the client, then we ought to get it right. If it is not used by the client (and I can think of no reason why it would be…?) then the field should be deleted from the output JSON.